File: src/pages/api/process.ts
import type { NextApiRequest, NextApiResponse } from "next";
import { google } from "googleapis";

// 設定
const EMAIL_COLUMN_PATTERNS = ["メールアドレス", "メアド", "eメール", "email", "e-mail", "mail"];

function getCredentials() {
  const credsJson = process.env.GOOGLE_CREDENTIALS;
  if (!credsJson) {
    throw new Error("GOOGLE_CREDENTIALS環境変数が設定されていません");
  }
  return JSON.parse(credsJson);
}

function extractSpreadsheetId(urlOrId: string): string {
  const patterns = [
    /https:\/\/docs\.google\.com\/spreadsheets\/d\/([a-zA-Z0-9_-]+)/,
    /^([a-zA-Z0-9_-]+)$/,
  ];
  for (const pattern of patterns) {
    const match = urlOrId.match(pattern);
    if (match) {
      return match[1];
    }
  }
  return urlOrId;
}

function normalizeEmail(email: string): string {
  if (!email) return "";
  return email.trim().normalize("NFKC").toLowerCase();
}

function findEmailColumn(headerRow: string[]): number {
  for (let i = 0; i < headerRow.length; i++) {
    const colLower = String(headerRow[i]).toLowerCase().trim();
    for (const pattern of EMAIL_COLUMN_PATTERNS) {
      if (colLower.includes(pattern.toLowerCase()) || pattern.toLowerCase().includes(colLower)) {
        return i;
      }
    }
  }
  return -1;
}

function colIndexToLetter(index: number): string {
  let result = "";
  let idx = index + 1;
  while (idx > 0) {
    const remainder = (idx - 1) % 26;
    result = String.fromCharCode(65 + remainder) + result;
    idx = Math.floor((idx - 1) / 26);
  }
  return result;
}

interface RequestBody {
  spreadsheet_url: string;
  email_route_map: Record<string, string[]>;
}

export default async function handler(
  req: NextApiRequest,
  res: NextApiResponse
) {
  if (req.method !== "POST") {
    return res.status(405).json({ success: false, error: "Method not allowed" });
  }

  try {
    const { spreadsheet_url, email_route_map } = req.body as RequestBody;

    if (!spreadsheet_url || !email_route_map) {
      return res.status(400).json({
        success: false,
        error: "スプレッドシートURLとメールアドレスマッピングが必要です",
      });
    }

    // スプレッドシートID抽出
    const spreadsheetId = extractSpreadsheetId(spreadsheet_url);

    // Google認証
    const credentials = getCredentials();
    const auth = new google.auth.GoogleAuth({
      credentials,
      scopes: ["https://www.googleapis.com/auth/spreadsheets"],
    });
    const sheets = google.sheets({ version: "v4", auth });

    // スプレッドシートデータ取得
    const response = await sheets.spreadsheets.values.get({
      spreadsheetId,
      range: "A:ZZ",
    });

    const allData = response.data.values;
    if (!allData || allData.length === 0) {
      return res.status(400).json({
        success: false,
        error: "スプレッドシートにデータがありません",
      });
    }

    // メールアドレス列を検出
    const headerRow = allData[0] as string[];
    const emailColIndex = findEmailColumn(headerRow);
    if (emailColIndex < 0) {
      return res.status(400).json({
        success: false,
        error: "メールアドレス列が見つかりません",
      });
    }

    // メールアドレスを抽出
    const emails: string[] = [];
    for (let i = 1; i < allData.length; i++) {
      const row = allData[i];
      if (row && emailColIndex < row.length && row[emailColIndex]) {
        emails.push(row[emailColIndex]);
      }
    }

    if (emails.length === 0) {
      return res.status(400).json({
        success: false,
        error: "メールアドレスが見つかりません",
      });
    }

    // 照合
    const results: { email: string; routes: string[]; found: boolean }[] = [];
    const notFoundEmails: string[] = [];

    for (const email of emails) {
      const normalized = normalizeEmail(email);
      const routes = email_route_map[normalized] || [];
      if (routes.length > 0) {
        results.push({ email, routes, found: true });
      } else {
        results.push({ email, routes: [], found: false });
        notFoundEmails.push(email);
      }
    }

    // UTAGE登録経路列の位置を決定
    let routeColIndex = headerRow.length;
    for (let i = 0; i < headerRow.length; i++) {
      if (headerRow[i] === "UTAGE登録経路") {
        routeColIndex = i;
        break;
      }
    }

    const routeColLetter = colIndexToLetter(routeColIndex);

    // 結果マッピング
    const resultMap: Record<string, string> = {};
    for (const r of results) {
      const normalized = normalizeEmail(r.email);
      resultMap[normalized] = r.routes.join(", ");
    }

    // データ準備
    const valuesToWrite: string[][] = [["UTAGE登録経路"]];
    for (let i = 1; i < allData.length; i++) {
      const row = allData[i];
      if (row && emailColIndex < row.length) {
        const email = row[emailColIndex];
        const normalized = normalizeEmail(email);
        const route = resultMap[normalized] || "";
        valuesToWrite.push([route]);
      } else {
        valuesToWrite.push([""]);
      }
    }

    // スプレッドシートに書き込み
    await sheets.spreadsheets.values.update({
      spreadsheetId,
      range: `${routeColLetter}1:${routeColLetter}${valuesToWrite.length}`,
      valueInputOption: "RAW",
      requestBody: {
        values: valuesToWrite,
      },
    });

    return res.status(200).json({
      success: true,
      total_count: emails.length,
      success_count: emails.length - notFoundEmails.length,
      not_found_count: notFoundEmails.length,
      not_found_emails: notFoundEmails.slice(0, 50),
    });

  } catch (error) {
    console.error("Error:", error);
    return res.status(500).json({
      success: false,
      error: String(error),
    });
  }
}

File: api/process.py
"""
UTAGE登録経路取得API - Vercel Serverless Function
"""

import json
import os
import re
import tempfile
import unicodedata
from http.server import BaseHTTPRequestHandler

import gspread
import pandas as pd
from google.oauth2 import service_account


# 設定
SCOPES = ["https://www.googleapis.com/auth/spreadsheets"]
EMAIL_COLUMN_PATTERNS = ["メールアドレス", "メアド", "eメール", "email", "e-mail", "mail"]
UTAGE_EMAIL_COLUMN = "メールアドレス"
UTAGE_ROUTE_COLUMN = "登録経路"


def get_credentials():
    """環境変数からGoogle認証情報を取得"""
    creds_json = os.environ.get("GOOGLE_CREDENTIALS")
    if not creds_json:
        raise Exception("GOOGLE_CREDENTIALS環境変数が設定されていません")

    creds_info = json.loads(creds_json)
    return service_account.Credentials.from_service_account_info(creds_info, scopes=SCOPES)


def extract_spreadsheet_id(url_or_id: str) -> str:
    """スプレッドシートURLからIDを抽出"""
    patterns = [
        r"https://docs\.google\.com/spreadsheets/d/([a-zA-Z0-9_-]+)",
        r"^([a-zA-Z0-9_-]+)$",
    ]
    for pattern in patterns:
        match = re.search(pattern, url_or_id)
        if match:
            return match.group(1)
    return url_or_id


def normalize_email(email: str) -> str:
    """メールアドレスを正規化"""
    if not email:
        return ""
    result = email.strip()
    result = unicodedata.normalize("NFKC", result)
    result = result.lower()
    return result


def find_email_column(header_row: list) -> int:
    """ヘッダー行からメールアドレス列を検出"""
    for i, col_name in enumerate(header_row):
        col_lower = str(col_name).lower().strip()
        for pattern in EMAIL_COLUMN_PATTERNS:
            if pattern.lower() in col_lower or col_lower in pattern.lower():
                return i
    return -1


def col_index_to_letter(index: int) -> str:
    """列インデックスをアルファベットに変換"""
    result = ""
    index += 1
    while index > 0:
        index, remainder = divmod(index - 1, 26)
        result = chr(65 + remainder) + result
    return result


def process_data(spreadsheet_url: str, csv_content: bytes) -> dict:
    """メイン処理"""
    # スプレッドシートID抽出
    spreadsheet_id = extract_spreadsheet_id(spreadsheet_url)

    # Google認証
    credentials = get_credentials()
    client = gspread.authorize(credentials)

    # スプレッドシートを開く
    spreadsheet = client.open_by_key(spreadsheet_id)
    worksheet = spreadsheet.sheet1

    # 全データ取得
    all_data = worksheet.get_all_values()
    if not all_data:
        raise Exception("スプレッドシートにデータがありません")

    # メールアドレス列を検出
    header_row = all_data[0]
    email_col_index = find_email_column(header_row)
    if email_col_index < 0:
        raise Exception("メールアドレス列が見つかりません")

    # メールアドレスを抽出
    emails = []
    for row in all_data[1:]:
        if email_col_index < len(row) and row[email_col_index]:
            emails.append(row[email_col_index])

    if not emails:
        raise Exception("メールアドレスが見つかりません")

    # CSVを読み込み
    with tempfile.NamedTemporaryFile(delete=False, suffix=".csv") as tmp:
        tmp.write(csv_content)
        tmp_path = tmp.name

    try:
        # エンコーディング判定
        try:
            csv_data = pd.read_csv(tmp_path, encoding="utf-8")
        except UnicodeDecodeError:
            csv_data = pd.read_csv(tmp_path, encoding="cp932")

        # 必須カラムチェック
        if UTAGE_EMAIL_COLUMN not in csv_data.columns:
            raise Exception(f"CSVに「{UTAGE_EMAIL_COLUMN}」列がありません")
        if UTAGE_ROUTE_COLUMN not in csv_data.columns:
            raise Exception(f"CSVに「{UTAGE_ROUTE_COLUMN}」列がありません")

        # インデックス作成
        email_to_routes = {}
        for _, row in csv_data.iterrows():
            email = str(row.get(UTAGE_EMAIL_COLUMN, ""))
            route = str(row.get(UTAGE_ROUTE_COLUMN, ""))
            if not email:
                continue
            normalized = normalize_email(email)
            if normalized not in email_to_routes:
                email_to_routes[normalized] = []
            if route not in email_to_routes[normalized]:
                email_to_routes[normalized].append(route)

        # 照合
        results = []
        not_found_emails = []
        for email in emails:
            normalized = normalize_email(email)
            routes = email_to_routes.get(normalized, [])
            if routes:
                results.append({"email": email, "routes": routes, "found": True})
            else:
                results.append({"email": email, "routes": [], "found": False})
                not_found_emails.append(email)

        # スプレッドシートに書き込み
        # UTAGE登録経路列の位置を決定
        existing_route_col = None
        for i, col_name in enumerate(header_row):
            if col_name == "UTAGE登録経路":
                existing_route_col = i
                break

        if existing_route_col is not None:
            route_col_index = existing_route_col
        else:
            route_col_index = len(header_row)

        route_col_letter = col_index_to_letter(route_col_index)

        # ヘッダー書き込み
        worksheet.update(f"{route_col_letter}1", [["UTAGE登録経路"]])

        # 結果マッピング
        result_map = {}
        for r in results:
            normalized = normalize_email(r["email"])
            result_map[normalized] = ", ".join(r["routes"])

        # データ書き込み
        values_to_write = []
        for row in all_data[1:]:
            if email_col_index < len(row):
                email = row[email_col_index]
                normalized = normalize_email(email)
                route = result_map.get(normalized, "")
                values_to_write.append([route])
            else:
                values_to_write.append([""])

        if values_to_write:
            cell_range = f"{route_col_letter}2:{route_col_letter}{len(values_to_write) + 1}"
            worksheet.update(cell_range, values_to_write)

        return {
            "success": True,
            "total_count": len(emails),
            "success_count": len(emails) - len(not_found_emails),
            "not_found_count": len(not_found_emails),
            "not_found_emails": not_found_emails[:50],  # 最大50件
        }

    finally:
        os.unlink(tmp_path)


class handler(BaseHTTPRequestHandler):
    def send_cors_headers(self):
        """CORSヘッダーを送信"""
        self.send_header("Access-Control-Allow-Origin", "*")
        self.send_header("Access-Control-Allow-Methods", "POST, OPTIONS")
        self.send_header("Access-Control-Allow-Headers", "Content-Type")

    def do_OPTIONS(self):
        """プリフライトリクエスト対応"""
        self.send_response(200)
        self.send_cors_headers()
        self.end_headers()

    def do_POST(self):
        try:
            # Content-Typeからboundaryを取得
            content_type = self.headers.get("Content-Type", "")

            # Content-Lengthを取得
            content_length = int(self.headers.get("Content-Length", 0))
            body = self.rfile.read(content_length)

            # multipart/form-dataをパース
            if "multipart/form-data" in content_type:
                boundary = content_type.split("boundary=")[1].encode()
                parts = body.split(b"--" + boundary)

                spreadsheet_url = None
                csv_content = None

                for part in parts:
                    if b'name="spreadsheet_url"' in part:
                        # テキストフィールド
                        lines = part.split(b"\r\n\r\n", 1)
                        if len(lines) > 1:
                            spreadsheet_url = lines[1].rstrip(b"\r\n--").decode("utf-8")
                    elif b'name="csv_file"' in part:
                        # ファイルフィールド
                        lines = part.split(b"\r\n\r\n", 1)
                        if len(lines) > 1:
                            csv_content = lines[1].rstrip(b"\r\n--")

                if not spreadsheet_url or not csv_content:
                    raise Exception("スプレッドシートURLとCSVファイルが必要です")

                result = process_data(spreadsheet_url, csv_content)

                self.send_response(200)
                self.send_header("Content-Type", "application/json")
                self.send_cors_headers()
                self.end_headers()
                self.wfile.write(json.dumps(result).encode())
            else:
                raise Exception("multipart/form-dataが必要です")

        except Exception as e:
            self.send_response(500)
            self.send_header("Content-Type", "application/json")
            self.send_cors_headers()
            self.end_headers()
            self.wfile.write(json.dumps({
                "success": False,
                "error": str(e)
            }).encode())
